File: client/lib/api/v2/schemas/logs.ts
/**
 * Logs V2 API Schemas
 * Schema definitions for logs v2 endpoints (read-only)
 */

import { z } from "zod";

// ============================================================================
// JSONB FIELD SCHEMAS
// ============================================================================

export const ActorDataSchema = z
  .object({
    userId: z.string().nullable().optional(),
    profileId: z.string().nullable().optional(),
    profileName: z.string().nullable().optional(),
  })
  .nullable();

export const SubjectDataSchema = z
  .object({
    entityId: z.string().nullable().optional(),
    entityType: z.string().nullable().optional(),
  })
  .nullable();

export const MetricsDataSchema = z
  .object({
    size: z.number().nullable().optional(),
    count: z.number().nullable().optional(),
    durationMs: z.number().nullable().optional(),
  })
  .nullable();

export const ContextDataSchema = z
  .object({
    route: z.string().nullable().optional(),
    function: z.string().nullable().optional(),
    component: z.string().nullable().optional(),
    provider: z.string().nullable().optional(),
    model: z.string().nullable().optional(),
  })
  .nullable();

export const ErrorDataSchema = z
  .object({
    code: z.string().nullable().optional(),
    name: z.string().nullable().optional(),
    stack: z.string().nullable().optional(),
    message: z.string().nullable().optional(),
  })
  .nullable();

export type ActorData = z.infer<typeof ActorDataSchema>;
export type SubjectData = z.infer<typeof SubjectDataSchema>;
export type MetricsData = z.infer<typeof MetricsDataSchema>;
export type ContextData = z.infer<typeof ContextDataSchema>;
export type ErrorData = z.infer<typeof ErrorDataSchema>;

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const LogsListRequestSchema = z.object({
  profileId: z.string(),
});

export type LogsListRequest = z.infer<typeof LogsListRequestSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Log item for list view
export const LogItemSchema = z.object({
  log_id: z.number(),
  event: z.string(),
  level: z.string(),
  message: z.string(),
  correlation_id: z.string(),
  actor: ActorDataSchema,
  subject: SubjectDataSchema,
  context: ContextDataSchema,
  error: ErrorDataSchema,
  created_at: z.string(),
  actor_name: z.string().nullable(),
});

export const LogsListResponseSchema = z.object({
  logs: z.array(LogItemSchema),
});

export type LogsListResponse = z.infer<typeof LogsListResponseSchema>;
export type LogItem = z.infer<typeof LogItemSchema>;

// ============================================================================
// CREATE LOG SCHEMAS
// ============================================================================

export const CorrelationDataSchema = z.object({
  correlationId: z.string().optional(),
  requestId: z.string().optional(),
  sessionId: z.string().optional(),
  attemptId: z.string().optional(),
  chatId: z.string().optional(),
});

export const CreateLogRequestSchema = z.object({
  event: z.string().default("legacy.message"),
  level: z.string().default("info"),
  message: z.string().default("Default Message"),
  correlation: CorrelationDataSchema.optional(),
  actor: z.record(z.any()).default({}),
  subject: z.record(z.any()).default({}),
  context: z.record(z.any()).default({}),
  error: z.record(z.any()).default({}),
});

export const CreateLogResponseSchema = z.object({
  success: z.boolean(),
  log_id: z.number().nullable().optional(),
});

export type CorrelationData = z.infer<typeof CorrelationDataSchema>;
export type CreateLogRequest = z.infer<typeof CreateLogRequestSchema>;
export type CreateLogResponse = z.infer<typeof CreateLogResponseSchema>;

// ============================================================================
// BULK DELETE SCHEMAS
// ============================================================================

export const BulkDeleteLogsRequestSchema = z.object({
  profileId: z.string(),
  ids: z.array(z.number()),
});

export const BulkDeleteLogsResponseSchema = z.object({
  success: z.boolean(),
  deleted_count: z.number(),
  message: z.string(),
});

export type BulkDeleteLogsRequest = z.infer<typeof BulkDeleteLogsRequestSchema>;
export type BulkDeleteLogsResponse = z.infer<
  typeof BulkDeleteLogsResponseSchema
>;

File: server/app/schemas/logs.py
"""Logs V2 API schemas."""

from typing import Any

from pydantic import BaseModel

# ============================================================================
# JSONB FIELD SCHEMAS
# ============================================================================


class ActorData(BaseModel):
    """Actor JSONB data."""

    userId: str | None = None
    profileId: str | None = None
    profileName: str | None = None


class SubjectData(BaseModel):
    """Subject JSONB data."""

    entityId: str | None = None
    entityType: str | None = None


class MetricsData(BaseModel):
    """Metrics JSONB data."""

    size: int | None = None
    count: int | None = None
    durationMs: int | None = None


class ContextData(BaseModel):
    """Context JSONB data."""

    route: str | None = None
    function: str | None = None
    component: str | None = None
    provider: str | None = None
    model: str | None = None


class ErrorData(BaseModel):
    """Error JSONB data."""

    code: str | None = None
    name: str | None = None
    stack: str | None = None
    message: str | None = None


# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class LogsListRequest(BaseModel):
    """Request for logs list."""

    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class LogItem(BaseModel):
    """Log item for list view."""

    log_id: int
    event: str
    level: str
    message: str
    correlation_id: str
    actor: ActorData
    subject: SubjectData
    context: ContextData
    error: ErrorData
    created_at: str
    actor_name: str | None


class LogsListResponse(BaseModel):
    """Response for logs list."""

    logs: list[LogItem]


# ============================================================================
# CREATE LOG SCHEMAS
# ============================================================================


class CorrelationData(BaseModel):
    """Correlation data for log creation."""

    correlationId: str | None = None
    requestId: str | None = None
    sessionId: str | None = None
    attemptId: str | None = None
    chatId: str | None = None


class CreateLogRequest(BaseModel):
    """Request to create a log entry."""

    event: str = "legacy.message"
    level: str = "info"
    message: str = "Default Message"
    correlation: CorrelationData = CorrelationData()
    actor: dict[str, Any] = {}
    subject: dict[str, Any] = {}
    context: dict[str, Any] = {}
    error: dict[str, Any] = {}


class CreateLogResponse(BaseModel):
    """Response from creating a log entry."""

    success: bool
    log_id: int | None = None


# ============================================================================
# BULK DELETE LOG SCHEMAS
# ============================================================================


class BulkDeleteLogsRequest(BaseModel):
    """Request to bulk delete logs."""

    profileId: str
    ids: list[int]


class BulkDeleteLogsResponse(BaseModel):
    """Response for bulk delete logs."""

    success: bool
    deleted_count: int
    message: str
